### DISCUSSION TITLE
Close connection issue in re-connection logic

### DISCUSSION THREAD

**Discussion by mithunpm** [gopcua/opcua#670] [Q&A]
Hi ,

We are attempting to use the reconnection module within the client, but with the current logic we are always encountering infinite loops and making a large number of connections within the server. While analyzing the problem 

client.go -   ActivateSessionWithContext mehod.

```go
return sc.SendRequestWithContext(ctx, req, s.resp.AuthenticationToken, func(v interface{}) error {
		var res *ua.ActivateSessionResponse
		if err := safeAssign(v, &res); err != nil {
			return err
		}

		// save the nonce for the next request
		s.serverNonce = res.ServerNonce

		// close the previous session
		//
		// https://github.com/gopcua/opcua/issues/474
		//
		// We decided not to check the error of CloseSession() since we
		// can't do much about it anyway and it creates a race in the
		// re-connection logic.
		c.CloseSessionWithContext(ctx) 

		c.setSession(s)
		return nil
	}) 
```
Between reconnections, the session will be terminated by calling c.CloseSessionWithContext(ctx). and perform an endless loop. 

Also while we disconnect from the server in between due to any network issue additional connection will be added on the server side.

Could you please give as some direction to move forward ?
@magiconair @kung-foo 

Thanks,
Mithun

**Comment by magiconair** [comment]
Is the question whether or not to close the session on error? What behavior would you expect?

**Reply by mithunpm** [reply]
Question is regarding the retry mechanism , How we can use the SDK retry mechanism  to resume the operations. Now if we given auto connect is true, this will go  to infinite loop and create infinite connection also. 

We are getting this issue when we encounter StatusBadEncodingLimitsExceeded error in between streaming ?  

Do we have any recommendation for this?
